File: backend/src/services/formulario.service.js
import supabase from '../config/supabase.js';

function validar(payload) {
  const { categoria, satisfaccion } = payload || {};
  if (!categoria) throw new Error('categoria requerida');
  if (typeof satisfaccion !== 'number') throw new Error('satisfaccion requerida (1..5)');
}

export async function insertarRespuesta(payload) {
  validar(payload);
  const {
    usuario_id,
    edad,
    genero,            // 'Masculino' | 'Femenino' | 'Otro'
    ciudad,
    categoria,         // 'Docentes' | 'Infraestructura' | 'Servicios' | 'Administración'
    satisfaccion,      // 1..5
    calidad_servicio,  // 1..5
    recomendar,        // 'Sí' | 'No'
    comentario
  } = payload;

  const { error } = await supabase.from('respuestas_formulario').insert({
    usuario_id: usuario_id || null,
    edad: edad ?? null,
    genero: genero || null,
    ciudad: ciudad || null,
    categoria,
    satisfaccion,
    calidad_servicio: calidad_servicio ?? null,
    recomendar: recomendar || null,
    comentario: comentario || null
  });

  if (error) throw error;
}

File: backend_dart/lib/services/formulario_service.dart
import 'package:supabase/supabase.dart';
import '../config/supabase.dart';

class FormularioInput {
  final String? usuarioId;
  final int? edad;
  final String? genero;
  final String? ciudad;
  final String categoria;        // requerido
  final int satisfaccion;        // 1..5 requerido
  final int? calidadServicio;
  final String? recomendar;
  final String? comentario;

  FormularioInput({
    this.usuarioId,
    this.edad,
    this.genero,
    this.ciudad,
    required this.categoria,
    required this.satisfaccion,
    this.calidadServicio,
    this.recomendar,
    this.comentario,
  });

  factory FormularioInput.fromMap(Map<String, dynamic> m) {
    void req(bool cond, String msg) { if (!cond) throw ArgumentError(msg); }
    req(m['categoria'] != null, 'categoria requerida');
    req(m['satisfaccion'] is int, 'satisfaccion requerida (int 1..5)');
    return FormularioInput(
      usuarioId: m['usuario_id'],
      edad: m['edad'],
      genero: m['genero'],
      ciudad: m['ciudad'],
      categoria: m['categoria'],
      satisfaccion: m['satisfaccion'],
      calidadServicio: m['calidad_servicio'],
      recomendar: m['recomendar'],
      comentario: m['comentario'],
    );
  }

  Map<String, dynamic> toRow() => {
    'usuario_id': usuarioId,
    'edad': edad,
    'genero': genero,
    'ciudad': ciudad,
    'categoria': categoria,
    'satisfaccion': satisfaccion,
    'calidad_servicio': calidadServicio,
    'recomendar': recomendar,
    'comentario': comentario,
  };
}

Future<void> insertarRespuesta(FormularioInput input) async {
  final res = await supabase.from('respuestas_formulario').insert(input.toRow());
  final error = (res as PostgrestResponse).error;
  if (error != null) throw error;
}
